add split-up spot detection notebooks (#38)
napari/napari-workshop-template: napari-workshops/notebooks/spot_detection_basic.md

```
For example, for better contrast you could choose an inverted colormap and minimum blending.

```{code-cell} ipython3
viewer.layers['nuclei'].colormap = 'I Forest'
viewer.layers['nuclei'].blending = 'minimum'
viewer.layers['spots'].colormap = 'I Orange'
viewer.layers['spots'].blending = 'minimum'
```

Once you are satisfied, you can print the output of any manual changes and then take a screenshot of the viewer.

```{code-cell} ipython3
# example of printing the nuclei layer visualization params
print('Colormap: ', viewer.layers['nuclei'].colormap.name)
print('Contrast limits: ', viewer.layers['nuclei'].contrast_limits)
print('Opacity: ', viewer.layers['nuclei'].opacity)
print('Blending: ', viewer.layers['nuclei'].blending)
```

```{code-cell} ipython3
nbscreenshot(viewer)
```

## Create an image filter

If you look carefully at the `spots` layer, you will notice that it contains background and
autofluorescence from the cells. It may help to just look at the single channel.

```{code-cell} ipython3
viewer.layers['nuclei'].visible = False
```

To improve spot detection, we will apply a high pass filter to improve the contrast of the spots.

A simple way to achieve this is to:

> 1. blur the image with a Gaussian, which removes high-frequency information—this is why we use it for denoising.

Author: ```suggestion
1. blur the image with a Gaussian, which removes high-frequency information (small features, including
our spots)—this is why we use it for denoising.
```

```

After loading the data, inspect it in the viewer and adjust the
layer settings to your liking (e.g., contrast limits, colormap). 
You can pan/zoom around the image by click/dragging to pan and scrolling with your
mousewheel or trackpad to zoom.

```{tip}
You can adjust a layer's opacity to see the change how much you see of the
layers that are "under" it.
```
For example, for better contrast you could choose an inverted colormap and minimum blending.

```{code-cell} ipython3
viewer.layers['nuclei'].colormap = 'I Forest'
viewer.layers['nuclei'].blending = 'minimum'
viewer.layers['spots'].colormap = 'I Orange'
viewer.layers['spots'].blending = 'minimum'
```

Once you are satisfied, you can print the output of any manual changes and then take a screenshot of the viewer.

```{code-cell} ipython3
# example of printing the nuclei layer visualization params
print('Colormap: ', viewer.layers['nuclei'].colormap.name)
print('Contrast limits: ', viewer.layers['nuclei'].contrast_limits)
print('Opacity: ', viewer.layers['nuclei'].opacity)
print('Blending: ', viewer.layers['nuclei'].blending)
```

```{code-cell} ipython3
nbscreenshot(viewer)
```

## Create an image filter

> If you look carefull at the `spots` layer, you will notice that it contains background and

Reviewer: ```suggestion
If you look carefully at the `spots` layer, you will notice that it contains background and
```

autofluorescence from the cells. It may help to just look at the single channel.

```{code-cell} ipython3
viewer.layers['nuclei'].visible = False
```

To improve spot detection, we will apply a high pass filter to improve the contrast of the spots.

A simple way to achieve this is to:

> 1. blur the image with a Gaussian, which removes high-frequency information—this is why we use it for denoising.

Reviewer: just fyi, I only understood the meaning of "high pass filter" and "high frequency information" very late in my image analysis journey. So it may be useful to define it, or to use simpler terms like, "which blurs out small features — including our spots — and leaves intact only the larger features, including background intensity"

2. subtract the blurred image from the original.

Lets try this using the `gaussian_filter` from `scipy` with a sigma of 2 px and lets clip any negative values:

```{code-cell} ipython3
import numpy as np
from scipy import ndimage as ndi

low_pass = ndi.gaussian_filter(spots, 2)
high_passed_spots = (spots - low_pass).clip(0)
```

Let's visualize both versions of the data.

```{code-cell} ipython3
viewer.add_image(low_pass, colormap="I Forest", blending="minimum")
viewer.add_image(high_passed_spots, colormap="I Blue", blending="minimum")
```

Toggle the visibility of the 3 spots layers to get a better sense of the effect of our filter.
Let's hide the `low_pass` and take a screenshot for documentation.

```{code-cell} ipython3
viewer.layers['low_pass'].visible = False
```

```{code-cell} ipython3
nbscreenshot(viewer)
```

## Detect spots

> Next, we will use one of the blob detection algorithms from sci-kit image to

Reviewer: ```suggestion
Next, we will use one of the blob detection algorithms from scikit-image to
```

perform the blob detection.

```{tip}
- See the [blob detection tutorial from scikit-image](https://scikit-image.org/docs/stable/auto_examples/features_detection/plot_blob.html). We recommend the [blob_log detector](https://scikit-image.org/docs/stable/api/skimage.feature.html#skimage.feature.blob_log), but feel free to experiment!
- See the "Note" from the blob_log docs: "The radius of each blob is approximately $\sqrt{2}\sigma$ for a 2-D image"
```

```{code-cell} ipython3
from skimage.feature import blob_log

# detect the spots on the filtered image
blobs_log = blob_log(
    high_passed_spots,
    max_sigma=3,

>     threshold=None, # use a relative threshold instead

Reviewer: ```suggestion
    threshold=None,  # use a relative threshold instead
```

sorrynotsorry 😜

Let's zoom in for a better look. You can explore interactively in the viewer and then you can explicitly
set the center of the field of view and the zoom factor.

```{code-cell} ipython3
viewer.camera.center = (200, 270)
viewer.camera.zoom = 8
```

```{code-cell} ipython3
nbscreenshot(viewer)
```

## Optional: using Points layer `properties`

`properties` is a table associated with a Points layer that can store additional data associated with
a data point. It has one row per data element (Point) and one column per feature. Importantly,
napari can not only display the values of associated features in the status bar, but also use them for
styling, e.g. for `face_color`. For more information, see the [Points layer guide](https://napari.org/stable/howtos/layers/points.html#setting-point-edge-and-face-color-with-properties).

```{note}
napari plans to deprecate `properties` in favor of `features` in a future release, but for the time being
the documentation for `properties` is superior, so we use that here. Note that some examples in the Gallery
have been updated to use `features`, but for the purposes of these simple examples you can replace
`features` with `properties` or vice versa.

> ```

Reviewer: This is pretty oof. 😣 Is there anything more than a find/replace on the docs to fix this? I could insta-merge and then we could use features here and get rid of this warning...

Author: This comment might be out-dated. I've not done full due-diligence. I think the example and tutorial are migrated by Andy.

Author: OK, I've moved everything to using features and changed the copy here, removing the note.
